tiledb/cc/group.cc

```diff
@@ -14,45 +14,76 @@ using namespace tiledb;
 using namespace tiledbpy::common;
 namespace py = pybind11;
 
-void put_metadata(Group &group, const std::string &key, py::array value) {
-  tiledb_datatype_t value_type = np_to_tdb_dtype(value.dtype());
+void put_metadata_numpy(Group &group, const std::string &key, py::array value) {
+  tiledb_datatype_t value_type;
+  try {
+    value_type = np_to_tdb_dtype(value.dtype());
+  } catch (const TileDBPyError &e) {
+    throw py::type_error(e.what());
+  }
 
   if (is_tdb_str(value_type) && value.size() > 1)
-    TPY_ERROR_LOC("array/list of strings not supported");
+    throw py::type_error("array/list of strings not supported");
 
   py::buffer_info value_buffer = value.request();
   if (value_buffer.ndim != 1)
-    TPY_ERROR_LOC("Only 1D Numpy arrays can be stored as metadata");
+    throw py::type_error("Only 1D Numpy arrays can be stored as metadata");
 
   py::size_t ncells = get_ncells(value.dtype());
   if (ncells != 1)
-    TPY_ERROR_LOC("Unsupported dtype for metadata");
+    throw py::type_error("Unsupported dtype for metadata");
 
   auto value_num = is_tdb_str(value_type) ? value.nbytes() : value.size();
-  group.put_metadata(key, value_type, value_num, value.data());
+  group.put_metadata(key, value_type, value_num,
+                     value_num > 0 ? value.data() : nullptr);
+}
+
+void put_metadata(Group &group, const std::string &key,
+                  tiledb_datatype_t value_type, uint32_t value_num,
+                  const char *value) {
+  group.put_metadata(key, value_type, value_num, value);
 }
 
 bool has_metadata(Group &group, const std::string &key) {
   tiledb_datatype_t _unused_value_type;
   return group.has_metadata(key, &_unused_value_type);
 }
 
-py::array get_metadata(Group &group, const std::string &key) {
+std::string get_key_from_index(Group &group, uint64_t index) {
+  std::string key;
+  tiledb_datatype_t tdb_type;
+  uint32_t value_num;
+  const void *value;
+
+  group.get_metadata_from_index(index, &key, &tdb_type, &value_num, &value);
+
+  return key;
+}
+
+py::tuple get_metadata(Group &group, const std::string &key) {
   tiledb_datatype_t tdb_type;
   uint32_t value_num;
-  const void *c_buf;
+  const void *value;
 
-  group.get_metadata(key, &tdb_type, &value_num, &c_buf);
+  group.get_metadata(key, &tdb_type, &value_num, &value);
 
   py::dtype value_type = tdb_to_np_dtype(tdb_type, 1);
 
-  if (tdb_type == TILEDB_STRING_UTF8)
-    value_num /= value_type.itemsize();
+  py::array py_buf;
+  if (value == nullptr) {
+    py_buf = py::array(value_type, 0);
+    return py::make_tuple(py_buf, tdb_type);
+  }
+
+  if (tdb_type == TILEDB_STRING_UTF8) {
+    value_type = py::dtype("|S1");
+  }
 
-  py::array py_buf(value_type, value_num);
-  memcpy(py_buf.mutable_data(), c_buf, py_buf.nbytes());
+  py_buf = py::array(value_type, value_num);
 
-  return py_buf;
+  memcpy(py_buf.mutable_data(), value, py_buf.nbytes());
```

Comment:
There is a `py::array` signature which copies from a pointer (by default, unless you pass a base object): https://github.com/pybind/pybind11/blob/68f80105007a574f114cfe899243bb3d96272a2e/include/pybind11/numpy.h#L745-L747

(usage/test: https://github.com/pybind/pybind11/blob/master/tests/test_numpy_array.cpp#L224-L230)